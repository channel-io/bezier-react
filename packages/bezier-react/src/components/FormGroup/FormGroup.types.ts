import type {
  AlphaBezierComponentProps,
  ChildrenProps,
} from '~/src/types/props'

import type { StackProps } from '~/src/components/Stack'

interface FormGroupProps extends
  AlphaBezierComponentProps,
  ChildrenProps,
  Partial<Pick<StackProps, 'direction' | 'spacing'>>,
  React.HTMLAttributes<HTMLDivElement> {}

export default FormGroupProps
