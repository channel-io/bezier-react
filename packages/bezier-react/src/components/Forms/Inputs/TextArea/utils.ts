import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
