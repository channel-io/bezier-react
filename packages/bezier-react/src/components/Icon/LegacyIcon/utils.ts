/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
