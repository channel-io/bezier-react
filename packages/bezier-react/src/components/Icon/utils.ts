import { type SVGProps } from 'react'

const BEZIER_ICON_SYMBOL = Symbol('bezier.icon')

export type IconSource = (props: SVGProps<SVGSVGElement>) => JSX.Element

export type BezierIcon = IconSource & { [BEZIER_ICON_SYMBOL]: true }

export function isBezierIcon(arg: unknown): arg is BezierIcon {
  return (
    typeof arg === 'function' &&
    arg[BEZIER_ICON_SYMBOL] === true
  )
}

export function createBezierIcon(source: IconSource): BezierIcon {
  source[BEZIER_ICON_SYMBOL] = true
  return source as BezierIcon
}
