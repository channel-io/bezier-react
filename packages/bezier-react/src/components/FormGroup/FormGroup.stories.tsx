import React from 'react'

import type {
  Meta,
  StoryFn,
  StoryObj,
} from '@storybook/react'

import { Checkbox } from '~/src/components/Checkbox'

import FormGroup from './FormGroup'
import type FormGroupProps from './FormGroup.types'

const meta: Meta<typeof FormGroup> = {
  component: FormGroup,
  argTypes: {
    spacing: {
      control: {
        type: 'number',
      },
    },
    direction: {
      control: {
        type: 'radio',
      },
      options: ['horizontal', 'vertical'],
    },
  },
}
export default meta

const Template: StoryFn<FormGroupProps> = (props) => (
  <FormGroup {...props}>
    <Checkbox>Option</Checkbox>
    <Checkbox>Option</Checkbox>
    <Checkbox>Option</Checkbox>
    <Checkbox>Option</Checkbox>
  </FormGroup>
)

export const Primary: StoryObj<FormGroupProps> = {
  render: Template,

  args: {
    spacing: 0,
    direction: 'vertical',
  },
}
