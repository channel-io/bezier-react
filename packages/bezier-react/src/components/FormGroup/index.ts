import FormGroup from './FormGroup'
import type FormGroupProps from './FormGroup.types'

export { FormGroup }
export type { FormGroupProps }
