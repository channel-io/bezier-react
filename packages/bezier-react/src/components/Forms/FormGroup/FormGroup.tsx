import React, { forwardRef } from 'react'

import useMergeRefs from '~/src/hooks/useMergeRefs'
import { noop } from '~/src/utils/functionUtils'

import { useFormControlContext } from '~/src/components/Forms/FormControl'

import type FormGroupProps from './FormGroup.types'

import * as Styled from './FormGroup.styled'

const FORM_GROUP_TEST_ID = 'bezier-react-form-group'

function FormGroup({
  testId = FORM_GROUP_TEST_ID,
  spacing = 6,
  direction = 'vertical',
  role = 'group',
  children,
  ...rest
}: FormGroupProps,
forwardedRef: React.Ref<HTMLDivElement>,
) {
  const contextValue = useFormControlContext()

  const {
    ref,
    ...ownProps
  } = contextValue?.getGroupProps(rest) ?? {
    ref: noop,
    ...rest,
  }

  const mergedRef = useMergeRefs(ref, forwardedRef)

  return (
    <Styled.Stack
      {...ownProps}
      data-testid={testId}
      ref={mergedRef}
      justify="start"
      align="stretch"
      spacing={spacing}
      direction={direction}
      role={role}
    >
      { children }
    </Styled.Stack>
  )
}

export default forwardRef(FormGroup)
