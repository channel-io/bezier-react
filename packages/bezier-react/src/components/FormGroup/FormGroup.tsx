import React, { forwardRef } from 'react'

import useMergeRefs from '~/src/hooks/useMergeRefs'
import { noop } from '~/src/utils/function'

import { useFormControlContext } from '~/src/components/FormControl'
import { Stack } from '~/src/components/Stack'

import { type FormGroupProps } from './FormGroup.types'

const FORM_GROUP_TEST_ID = 'bezier-form-group'

export const FormGroup = forwardRef<HTMLDivElement, FormGroupProps>(
  function FormGroup(
    { spacing = 6, direction = 'vertical', role = 'group', children, ...rest },
    forwardedRef
  ) {
    const contextValue = useFormControlContext()

    const { ref, ...ownProps } = contextValue?.getGroupProps(rest) ?? {
      ref: noop,
      ...rest,
    }

    const mergedRef = useMergeRefs(ref, forwardedRef)

    return (
      <Stack
        data-testid={FORM_GROUP_TEST_ID}
        ref={mergedRef}
        wrap
        justify="start"
        align="stretch"
        spacing={spacing}
        direction={direction}
        role={role}
        {...ownProps}
      >
        {children}
      </Stack>
    )
  }
)
