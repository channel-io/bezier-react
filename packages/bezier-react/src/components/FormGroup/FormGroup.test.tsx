import React from 'react'

import { render } from '~/src/utils/test'

import { FormGroup } from './FormGroup'
import { type FormGroupProps } from './FormGroup.types'

describe('FormGroup >', () => {
  let props: FormGroupProps

  beforeEach(() => {
    props = {
      spacing: 6,
      direction: 'vertical',
    }
  })

  const renderComponent = () => render(<FormGroup {...props} />)

  it('Snapshot >', () => {
    const { getByRole } = renderComponent()
    const rendered = getByRole('group')

    expect(rendered).toMatchSnapshot()
  })
})
