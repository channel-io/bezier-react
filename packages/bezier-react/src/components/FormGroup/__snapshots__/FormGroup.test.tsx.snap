// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGroup > Snapshot > 1`] = `
<div
  class="Stack display-flex direction-vertical justify-start align-stretch wrap margin layout"
  data-testid="bezier-react-form-group"
  role="group"
  style="--b-stack-spacing: 6px;"
/>
`;
